### ISSUE TITLE
0.13: [Bug]: Graph view on Live page doesn't update correctly

### ISSUE THREAD

**Issue by eysi09** [garden-io/garden#4597]
## Garden Bonsai (0.13) Bug

NOTE: Before reading the below, just know that I'm happy to hop on a Discord call to discuss in person. But this took awhile to uncover so I wanted to jot down all the context while it's still fresh in my mind and because some stakeholders are on vacation.

### Current Behavior

The graph view on the Live page doesn't update properly. 

This happens because Garden stops emitting events from a given command that was requested via websocket if another command is also started over websockets before the first one finishes.

I've done some digging around and the problem is pretty deep in the engine. I see what it is, is but I'm not entirely sure how to address it.

I'll list what I know below. 

### Problem

TL;DR: We lookup Garden instances with keys that don't have namespaces but write them with keys that do.   

I suppose there's a bit more to it than the TL;DR suggests. This is the entire flow:

1. **Command A** is requested over websocket. 
2. The instance manager checks whether a valid instance already exists [here](https://github.com/garden-io/garden/blob/master/core/src/server/instance-manager.ts#L134). The namespace is usually not available here so the instance key will be something like `env=dev`. 
3. A Garden instance is created and added to the instance map via `InstanceManager.set()`.
4. In the `set` method we again read the instance key but this time using the `getInstanceKey` method on the newly created Garden instance. See [here](https://github.com/garden-io/garden/blob/master/core/src/server/instance-manager.ts#L217).
5. The Garden instance actually knows the namespace so this time the key will be something like `env=dev-my-namespace`. This is the key we use when writing to the map.
6. We continue processing the command.
7. **Command B** is requested over websocket before **Command A** is done.
8. As in step 2 above, the instance manager checks whether a valid instance already exists. Again it uses the `env=dev` key whereas the instance for **Command A** which is indeed valid is stored via the `env=dev-my-namespace` key. 
9. Because it doesn't find the a valid instance, it again creates a new instance like in step 3 above. 
10. In the `set` method however, it now finds a valid, existing Garden instance because the instance key used there contains the namespace, like in step 4 above. 
11. [This code](https://github.com/garden-io/garden/blob/main/core/src/server/instance-manager.ts#L255-L263) then  unsubscribes the listener for **Command A** and re-subscribes it to the Garden instance for **Command B** which doesn't use the same `EventBus`. 
12. As a result, the `EventBus` for the instance belonging to **Command A** continues emitting but **Command A** is now subscribed to the bus for **Command B**. This means the events for **Command A** never reach Cloud. 

I know that's a lot of context, I'm only listing all of it because the appropriate solution could apply to almost any of the steps, depending on what we identify as the underlying problem. 

### Suggested solutions

Again, I'm not entirely sure what the right course of action is, but here are some things that could work, if only to paint a picture:

1. Resolve the namespace before ensuring the instance. That means we'd get the right key in step 2 above and the rest just works. However, the whole idea is to not resolve config at this point.
2. Only key instances by environment unless the namespace is specifically set via the command line. This also means we get the right key in step 2 above, but will not work if a user changes the namespace at the config level while the dev command is running.   
3. Skip unsubscribing the existing listener in step 11 above. This might cause memory leaks.
4. If there's an existing instance, copy the `EventBus` from that one to the new one. Works but we'd still end up creating a new Garden instance on every request.

**Comment by eysi09** [comment]
Setting `priority:high` because of the impact this has on the Live page.

**Comment by eysi09** [comment]
Maybe some combination of 2. and 3./4. work. 

That is, we key by environment unless namespace is set via command line. 

Now, if the config changes and thereby the namespace Garden would've resolved, we create a new instance (like we already do) but make sure any command still running for the old instance completes. Maybe by copying over the event bus or simply not unsubscribing the listener (or some other approach?).

**Comment by Orzelius** [comment]
@eysi09 you have all of the context here, would it  be possible for you to handle this yourself?
